### ISSUE TITLE
[DataGrid] Fix docs, add missing `onCellFocusOut`

### ISSUE THREAD

**Issue by missbruni** [mui-org/material-ui-x#2146]
On release https://github.com/mui-org/material-ui-x/releases/tag/v4.0.0-alpha.32 it seems onEditCellChangeCommitted is now not returning the event, on the release itself there seems to be no explanation of how to go around in case you do want to stop the default component behaviour for this event. 

- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Without being able to stop event propagation we are no longer able to stop the cell from going into view mode once commited with clicking out on a edit more custom row implementation. 

Using alpha 33
https://codesandbox.io/s/material-demo-forked-dqx3v

## Expected Behavior 🤔

Using alpha 31
https://codesandbox.io/s/material-demo-forked-fzthu

## Steps to Reproduce 🕹

- Go on demo links for expected or current scenarios
- Click on Edit for a row
- Edit a cell and click out (commit) to the next one

You will see on expected that the cell remains on edit mode and on current it doesnt. This is due to onEditCellChangeCommitted not having access to the event any longer. 

## Context 🔦

We are trying to remain on edit mode upon committing a cell change on clicking outside as we have a custom row implementation with a master save. If this is an intentional change, it would be appreciated if we could be given a workaround for this behaviour.

**Comment by m4theushw** [comment]
Clicking outside a cell does two things: switch back to view mode and commits the value. Stopping the propagation in `onEditCellChangeCommitted` only prevents the commiting. Thus, the row will not be updated, but the cell will still go to the view mode. To prevent it from reverting to view mode, cancel the default behavior of the outside click with the `onCellFocusOut`:

```jsx
<DataGrid
  onCellFocusOut={(params, event) => {
    event.defaultMuiPrevented = true;
  }}
/>
```

There's an example using this handler to also achieve this master save implementation: https://material-ui.com/components/data-grid/editing/#edit-using-external-button

---

@oliviertassinari The example in the docs I mentioned is labeled as a Pro feature but it's just a custom implementation using the props. What is the original intention of having it labeled like it? Is the ability to stop the propagation of events?

**Comment by dtassone** [comment]
@m4theushw the external button demo uses apiRef which is why it is marked as a pro demo. Stopping the propagation of events works in MIT as well

**Comment by missbruni** [comment]
@m4theushw, what @dtassone said  and the same goes for the onCellFocusOut, it is not available on the MIT DataGrid props.

**Comment by missbruni** [comment]
> Clicking outside a cell does two things: switch back to view mode and commits the value. Stopping the propagation in `onEditCellChangeCommitted` only prevents the commiting. Thus, the row will not be updated, but the cell will still go to the view mode. 

Also, just clarifying on the above when you say "but the cell will still go to the view mode. ", with the stopPropagation on version 31 and below, the cell does not go to view mode as you can see https://codesandbox.io/s/material-demo-forked-fzthu.

**Comment by m4theushw** [comment]
The [DataGrid docs page](https://material-ui.com/api/data-grid/data-grid/) is missing this prop, we need to update it. The only Pro feature in the example is the access to the API.

There are two actions on this issue:

- Allow to stop the propagation of the `cellEditExit` event. This one is a better option, because `cellFocusOut` is too broad.
- Fix the docs to include the missing prop.

**Comment by missbruni** [comment]
Is also not on the definition file for data-grid.d.ts

**Comment by m4theushw** [comment]
> Also, just clarifying on the above when you say "but the cell will still go to the view mode. ", with the stopPropagation on version 31 and below, the cell does not go to view mode as you can see https://codesandbox.io/s/material-demo-forked-fzthu.

This occurs because clicking outside the cell there's no event to stop the propagation. You can use the trick of https://github.com/mui-org/material-ui-x/issues/2146#issuecomment-879922184 and also stop the propagation in the `onEditCellChangeCommitted` prop. In #1403 we want to improve the event system to allow to stop the propagation of any event.

https://github.com/mui-org/material-ui-x/blob/a316c76d427393547a0778f048543214ddf2cf6e/packages/grid/_modules_/grid/hooks/features/rows/useGridEditRows.ts#L58

**Comment by missbruni** [comment]
oops, I miss that. I think I was looking at 31. Will try it out. Thanks!

**Comment by oliviertassinari** [comment]
@m4theushw Great work on this issue 👍 . I have changed a bit the wording of stop the propagation of the event to stop the internal handling to avoid confusion. It's closer to the result we are looking (having the event bubble up or not is irrelevant for us)

**Comment by m4theushw** [comment]
> I assume the same event is shared between the two, so if we want to isolate the two, we would need to get the call order right, (execute all the handlers of cellFocusOut before going to cellEditExit). In practice, it's brittle.

@oliviertassinari Yes, it's shared. The user shouldn't have to know the order which they are called. We should allow to prevent the default behavior of one, without affecting another one.

**Comment by oliviertassinari** [comment]
@m4theushw I guess we could create some sort of an attraction to reset `defaultMuiPrevented` once the event is done being processed.

**Comment by m4theushw** [comment]
Point 1 in https://github.com/mui-org/material-ui-x/issues/2146#issuecomment-879968581 was fixed by #2179
